### PR TITLE
Add OTA update

### LINKED ISSUE #10: Implement `io.edgehog.devicemanager.OTARequest` and `OTAResponse` Interfaces in Device Runtime
See also: https://github.com/edgehog-device-manager/edgehog-astarte-interfaces

### PR DESCRIPTION
**Pull request by shikoku1**
Implement `io.edgehog.devicemanager.OTARequest` and `io.edgehog.devicemanager.OTAResponse`

D-Bus code generation is not automated due to a zbus issue [1], a manual implementation is being used for now

Closes #10

[1] https://gitlab.freedesktop.org/dbus/zbus/-/issues/168

### CODE REVIEW

File: src/updatemanager.rs [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+pub enum OtaError {
+    #[error("Bundle download failed")]
+    Network,
+    #[error("Deployment failed")]
+    Deploy,
+    #[error("OTA failed")]
+    Failed,
+}
+
+pub struct UpdateManager<'a> {
```
Comment (harlem88) [inline src/updatemanager.rs:44]: @bettio What do you think about this name? IMHO it should be called something like `OTAHandler`

File: src/error.rs [hunk 2]
```diff
@@ -34,6 +34,21 @@ pub enum DeviceManagerError {
     #[error(transparent)]
     IOError(#[from] std::io::Error),
 
+    #[error(transparent)]
+    ZbusError(#[from] zbus::Error),
+
+    #[error("update error")]
+    UpdateError(String),
+
     #[error("urrecoverable error")]
```
Comment (harlem88) [inline src/error.rs:43]: Change this to `unrecoverable`

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by FrancescoVaiani — APPROVED** [review]

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/updatemanager.rs [hunk 3]
```diff
@@ -0,0 +1,354 @@
+/*
```
Comment (harlem88) [inline src/updatemanager.rs:1]: Chage updatemanager.rs in `ota_handler.rs`

File: src/updatemanager.rs [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+#[derive(Debug)]
+enum OtaStatus {
+    InProgress,
+    Done,
+    Error(OtaError),
+}
+
+pub struct UpdateManager<'a> {
```
Comment (harlem88) [inline src/updatemanager.rs:54]: Change UpdateManager in `OTAHandler`

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 5]
```diff
@@ -46,10 +53,30 @@ impl DeviceManager {
         )
         .interface_directory(&opts.interface_json_path)?
         .build();
+        info!("Starting");
 
         let device = astarte_sdk::AstarteSdk::new(&sdk_options).await?;
 
-        Ok(Self { sdk: device })
+        let mut updmanager = ota_handler::OTAHandler::new(&opts).await?;
```
Comment (harlem88) [inline src/lib.rs:60]: Change the local variable name to `ota_handler`

File: src/ota_handler.rs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            state_file_path: opts.state_file_path.clone(),
+            download_file_path: opts.download_file_path.clone(),
+        })
+    }
+
+    pub async fn last_error(&self) -> Result<String, DeviceManagerError> {
+        Ok(self.rauc.last_error().await?)
+    }
+
+    pub async fn update_event(
```
Comment (harlem88) [inline src/ota_handler.rs:82]: Change this to `ota_event`

File: src/ota_handler.rs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+                            .ok(),
+                    };
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_update_event(
```
Comment (harlem88) [inline src/ota_handler.rs:122]: Change this to `handle_ota_event`

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

File: src/ota_handler.rs [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+    /// Deployment failed
+    #[error("OTAErrorDeploy")]
+    Deploy,
+    /// OTA failed
+    #[error("OTAFailed")]
+    Failed,
+}
+
+#[derive(Debug)]
+enum OtaStatus {
```
Comment (harlem88) [inline src/ota_handler.rs:48]: Change this to`OTAStatus`

File: src/ota_handler.rs [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+        if output.status.success() && output.stderr.is_empty() {
+            panic!("Reboot command was successful, bye");
+        } else {
+            error!("Reboot failed {:?}", output.stderr);
+        }
+
+        Ok(())
+    }
+
+    pub async fn check_pending_otaresponse(
```
Comment (harlem88) [inline src/ota_handler.rs:220]: Change this to `check_pending_ota_response`

File: src/ota_handler.rs [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+                    .await?;
+            }
+
+            std::fs::remove_file(&self.state_file_path)?;
+        }
+
+        Ok(())
+    }
+
+    fn map_otastatus(status: OtaStatus) -> (String, String) {
```
Comment (harlem88) [inline src/ota_handler.rs:244]: Change this to `map_ota_status`

File: src/ota_handler.rs [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+    Err(OtaError::Network.into())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::ota_handler::{OTAHandler, OtaError, OtaStatus};
+
+    #[test]
+    fn otastatus() {
```
Comment (harlem88) [inline src/ota_handler.rs:330]: Change this to `ota_status`

**Review by harlem88 — CHANGES_REQUESTED** [review]

File: src/error.rs [hunk 12]
```diff
@@ -34,6 +34,21 @@ pub enum DeviceManagerError {
...
     FatalError(String),
+
+    #[error(transparent)]
+    ReqwestError(#[from] reqwest::Error),
+
+    #[error(transparent)]
+    SerdeJsonError(#[from] serde_json::Error),
+
+    #[error(transparent)]
+    OtaError(#[from] crate::ota_handler::OTAError),
```
Comment (harlem88) [inline src/error.rs:53]: Change also this to `OTAError`

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — APPROVED** [review]

File: src/ota_handler.rs [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+        if output.status.success() && output.stderr.is_empty() {
+            panic!("Reboot command was successful, bye");
+        } else {
+            error!("Reboot failed {:?}", output.stderr);
+        }
+
+        Ok(())
+    }
+
+    pub async fn check_pending_ota_response(
```
Comment (Annopaolo) [inline src/ota_handler.rs:220]: `check_pending_ota_response` is a bit unclear on side effects. Maybe `ensure_pending_ota_response()?

File: src/ota_handler.rs [hunk 14]
```diff
@@ -0,0 +1,352 @@
...
+                    .await?;
+            }
+
+            std::fs::remove_file(&self.state_file_path)?;
+        }
+
+        Ok(())
+    }
+
+    fn map_ota_status(status: OTAStatus) -> (String, String) {
```
Comment (Annopaolo) [inline src/ota_handler.rs:244]: `map` is too generic, maybe a better name can be used also here

**Review by Annopaolo — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 15]
```diff
@@ -53,7 +53,9 @@ async fn main() -> Result<(), edgehog_device_runtime::error::DeviceManagerError>
         device_id,
         credentials_secret,
         pairing_url,
-        interface_json_path: "./interfaces".to_string(),
+        interface_json_path: "/tmp/interfaces".to_string(),
```
Comment (bettio) [inline src/main.rs:56]: revert previous line.

File: src/main.rs [hunk 16]
```diff
@@ -53,7 +53,9 @@ async fn main() -> Result<(), edgehog_device_runtime::error::DeviceManagerError>
...
+        state_file_path: "/data/dm-update".to_string(),
```
Comment (bettio) [inline src/main.rs:57]: let's add: TODO: change following paths

**Review by bettio — CHANGES_REQUESTED** [review]

**Review by bettio — APPROVED** [review]

**Review by bettio — APPROVED** [review]

**Pull request merged**